bundles/org.openhab.binding.livisismarthome/src/main/java/org/openhab/binding/livisismarthome/internal/handler/LivisiBridgeHandler.java

```diff
@@ -0,0 +1,926 @@
+                        break;
+                    case BaseEventDTO.TYPE_NEW_MESSAGE_RECEIVED:
+                    case BaseEventDTO.TYPE_MESSAGE_CREATED:
+                        final Optional<MessageEventDTO> messageEvent = gson.fromJson(msg, MessageEventDTO.class);
+                        if (messageEvent.isPresent()) {
+                            handleNewMessageReceivedEvent(Objects.requireNonNull(messageEvent.get()));
+                        }
+                        break;
+                    case BaseEventDTO.TYPE_MESSAGE_DELETED:
+                        handleMessageDeletedEvent(event);
+                        break;
+                    default:
+                        logger.debug("Unsupported event type {}.", event.getType());
+                        break;
+                }
+            }
+        } catch (IOException | RuntimeException e) {
+            logger.debug("Error with Event: {}", e.getMessage(), e);
+            handleClientException(e);
+        }
+    }
+
+    @Override
+    public void onError(final Throwable cause) {
+        if (cause instanceof Exception) {
+            handleClientException((Exception) cause);
+        }
+    }
+
+    /**
+     * Handles the event that occurs, when the state of a device (like reachability) or a capability (like a temperature
+     * value) has changed.
+     *
+     * @param event event
+     */
+    private void handleStateChangedEvent(final EventDTO event) throws IOException {
+
+        // CAPABILITY
+        if (event.isLinkedtoCapability()) {
+            logger.trace("Event is linked to capability");
+            final Optional<DeviceDTO> device = deviceStructMan.getDeviceByCapabilityId(event.getSourceId());
+            notifyDeviceStatusListeners(device, event);
+
+            // DEVICE
+        } else if (event.isLinkedtoDevice()) {
+            logger.trace("Event is linked to device");
+            final String sourceId = event.getSourceId();
+
+            final Optional<DeviceDTO> bridgeDevice = deviceStructMan.getBridgeDevice();
+            final Optional<DeviceDTO> device;
+            if (bridgeDevice.isPresent() && !sourceId.equals(bridgeDevice.get().getId())) {
+                device = deviceStructMan.refreshDevice(sourceId, isSHCClassic());
+            } else {
+                device = deviceStructMan.getDeviceById(sourceId);
+            }
+            notifyDeviceStatusListeners(device, event);
+
+        } else {
+            logger.debug("link type {} not supported (yet?)", event.getSourceLinkType());
+        }
+    }
+
+    /**
+     * Handles the event that occurs, when the connectivity of the bridge has changed.
+     *
+     * @param event event
+     */
+    private void handleControllerConnectivityChangedEvent(final EventDTO event) throws IOException {
+
+        final Boolean connected = event.getIsConnected();
+        if (connected != null) {
+            final ThingStatus thingStatus = createThingStatus(connected);
```

Comment:
Now reading this one more time, one doubt: When connected is false, what does it exactly mean? openHAB can communicate with the controller, but the controller is somehow not connected? The reason I'm asking is that `updateStatus` in line 607 simply sets status **OFFLINE** without any `ThingStatusDetail`. Perhaps it would be nice for the user to know if problem is between openHAB and controller, or something within the controller?

Reply:
now better (with COMMUNICATION_ERROR)?

Reply:
I think it depends on its meaning: What does it exactly mean that connected is false?

Reply:
The reason I keep asking about this is that user will only see **OFFLINE**/**COMMUNICATION_ERROR** without any additional text (third parameter for `updateStatus`). So I'm wondering if this could be improved, but first I would need to understand the issue, which boils down to:
- Did openHAB lose the connection to the LIVISI controller?
- Or did the LIVISI controller lose the connection to the device?

Reply:
As an example, this is how the Hue binding marks bulbs offline if they are not reachable by the bridge:
```
updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "@text/offline.light-not-reachable");
```

This would be similar to the case where the LIVISI controller has lost connection to the device.

Reply:
In the Innogy-world (with BackEnd-Communication) COMMUNICATION_ERROR at the Bridge-Thing means, that the SHC (Bridge) can not communicate with the back end. COMMUNICATION_ERROR at a Device-Thing means, that SHC cannot reach the Device.

In the new LIVISI-World, there is no BackEnd anymore. So the COMMUNICTAION_ERROR at the Bridge-Thing should occur when OpenHAB cannot reach the SHC. The behaviour at the Device-Thing should be the same as before in the innogy-world.

Reply:
@RalphSester - thank you. So in that case I would suggest to handle it the same way as in the Hue binding (and deCONZ also): Do not set COMMUNICATION_ERROR, but instead supply a message describing that the device is not reachable.

Reply:
solved

Reply:
Sorry about previous comment about "device", as this is the bridge handler. Which leads me to one more question (sorry): openHAB is communicating with the SHC, right? And this communication is working, since we received an Event DTO. So what does in mean for the bridge that `isConnected` is false? I understand it for a device context, but how can SHC itself be unreachable/not connected, when we are communicating with it?

Reply:
@Novanic - can you help clarifying this for me as this is last open comment?

Reply:
I was also wondering about it. But I asked the vendor and the event still exists (technically). It means that the SHC lost the connection to the backend / Internet. That shouldn't happen anymore with the local API because there is no backend connection anymore but I know the SHC currently requires still an Internet connection to synchronize the time with the time servers. So the event still exists and it isn't 100% clear that we could safely remove the handling of it. Therefore I thought it is better to just improve the message ;-) instead of risking bugs short before the OpenHAB release.

Reply:
Perhaps you could test taking the controller offline from the internet? To me it sounds like it doesn't make sense to take the bridge **OFFLINE** when controller is not connected to the Internet. Instead you could leave it online and log an information that connection to Internet is lost. With current code you are losing the one of the big advantages of having a local API, since it will not work when you are offline.